File: app/text_utils.py
from __future__ import annotations

import re
from collections import Counter
from pathlib import Path


STOPWORDS = {
    "a",
    "an",
    "and",
    "are",
    "as",
    "at",
    "be",
    "by",
    "do",
    "does",
    "for",
    "from",
    "how",
    "i",
    "if",
    "in",
    "is",
    "it",
    "my",
    "of",
    "on",
    "or",
    "the",
    "to",
    "what",
    "when",
    "where",
    "who",
    "with",
    "you",
    "your",
}


CATEGORY_KEYWORDS: dict[str, set[str]] = {
    "hr": {"employee", "onboarding", "leave", "benefit", "attendance", "performance", "salary"},
    "finance": {"reimbursement", "expense", "invoice", "payment", "budget", "travel", "procurement", "approval"},
    "it": {"account", "vpn", "device", "permission", "access", "ticket", "software", "github"},
    "engineering": {"deploy", "deployment", "api", "code", "review", "database", "incident", "rollback", "repository"},
    "security": {"security", "credential", "password", "access", "permission", "audit", "compliance"},
    "legal": {"contract", "legal", "privacy", "terms", "compliance", "policy"},
}


def slugify(value: str) -> str:
    value = value.strip().lower()
    value = re.sub(r"[^a-z0-9\u4e00-\u9fff]+", "_", value)
    value = re.sub(r"_+", "_", value).strip("_")
    return value or "document"


def tokenize(text: str) -> list[str]:
    terms = re.findall(r"[\w\u4e00-\u9fff]+", text.lower())
    return [term for term in terms if len(term) > 1 and term not in STOPWORDS]


def top_terms(text: str, limit: int = 12) -> list[str]:
    counts = Counter(tokenize(text))
    return [term for term, _ in counts.most_common(limit)]


def title_from_markdown(path: Path, text: str) -> str:
    for line in text.splitlines():
        stripped = line.strip()
        if stripped.startswith("#"):
            return stripped.lstrip("#").strip() or path.stem.replace("_", " ").title()
    return path.stem.replace("_", " ").title()


def summarize_markdown(text: str, max_chars: int = 240) -> str:
    lines = []
    in_front_matter = False
    for idx, raw in enumerate(text.splitlines()):
        line = raw.strip()
        if idx == 0 and line == "---":
            in_front_matter = True
            continue
        if in_front_matter:
            if line == "---":
                in_front_matter = False
            continue
        if line and not line.startswith("#") and not line.startswith("|"):
            lines.append(re.sub(r"\s+", " ", line))
        if sum(len(line) for line in lines) > max_chars:
            break
    summary = " ".join(lines).strip()
    return summary[:max_chars].rstrip()


def score_text(query_terms: set[str], candidate_text: str) -> float:
    if not query_terms:
        return 0.0
    candidate_terms = set(tokenize(candidate_text))
    overlap = query_terms & candidate_terms
    return len(overlap) / len(query_terms)


def infer_categories(text: str, minimum: int = 1) -> list[str]:
    terms = set(tokenize(text))
    scores = []
    for category, keywords in CATEGORY_KEYWORDS.items():
        score = len(terms & keywords)
        if score:
            scores.append((score, category))
    scores.sort(reverse=True)
    categories = [category for _, category in scores]
    return categories or ["general"][:minimum]

File: src/textUtils.ts
import path from "node:path";

export const STOPWORDS = new Set([
  "a",
  "an",
  "and",
  "are",
  "as",
  "at",
  "be",
  "by",
  "do",
  "does",
  "for",
  "from",
  "how",
  "i",
  "if",
  "in",
  "is",
  "it",
  "my",
  "of",
  "on",
  "or",
  "the",
  "to",
  "what",
  "when",
  "where",
  "who",
  "with",
  "you",
  "your"
]);

export const CATEGORY_KEYWORDS: Record<string, Set<string>> = {
  hr: new Set(["employee", "onboarding", "leave", "benefit", "attendance", "performance", "salary"]),
  finance: new Set(["reimbursement", "expense", "invoice", "payment", "budget", "travel", "procurement", "approval"]),
  it: new Set(["account", "vpn", "device", "permission", "access", "ticket", "software", "github"]),
  engineering: new Set(["deploy", "deployment", "api", "code", "review", "database", "incident", "rollback", "repository"]),
  security: new Set(["security", "credential", "password", "access", "permission", "audit", "compliance"]),
  legal: new Set(["contract", "legal", "privacy", "terms", "compliance", "policy"])
};

export function slugify(value: string): string {
  const slug = value
    .trim()
    .toLowerCase()
    .replace(/[^a-z0-9\u4e00-\u9fff]+/gu, "_")
    .replace(/_+/g, "_")
    .replace(/^_+|_+$/g, "");
  return slug || "document";
}

export function tokenize(text: string): string[] {
  return [...text.toLowerCase().matchAll(/[\w\u4e00-\u9fff]+/gu)]
    .map((match) => match[0])
    .filter((term) => term.length > 1 && !STOPWORDS.has(term));
}

export function topTerms(text: string, limit = 12): string[] {
  const counts = new Map<string, number>();
  for (const term of tokenize(text)) counts.set(term, (counts.get(term) ?? 0) + 1);
  return [...counts.entries()]
    .sort((a, b) => b[1] - a[1])
    .slice(0, limit)
    .map(([term]) => term);
}

export function scoreText(queryTerms: Set<string>, candidateText: string): number {
  if (!queryTerms.size) return 0;
  const candidateTerms = new Set(tokenize(candidateText));
  let overlap = 0;
  for (const term of queryTerms) if (candidateTerms.has(term)) overlap += 1;
  return overlap / queryTerms.size;
}

export function titleFromMarkdown(filePath: string, text: string): string {
  for (const line of text.split(/\r?\n/)) {
    const stripped = line.trim();
    if (stripped.startsWith("#")) return stripped.replace(/^#+/, "").trim() || titleFromPath(filePath);
  }
  return titleFromPath(filePath);
}

export function summarizeMarkdown(text: string, maxChars = 240): string {
  const lines: string[] = [];
  let inFrontMatter = false;
  const rawLines = text.split(/\r?\n/);
  rawLines.forEach((raw, idx) => {
    const line = raw.trim();
    if (idx === 0 && line === "---") {
      inFrontMatter = true;
      return;
    }
    if (inFrontMatter) {
      if (line === "---") inFrontMatter = false;
      return;
    }
    if (line && !line.startsWith("#") && !line.startsWith("|") && lines.join("").length <= maxChars) {
      lines.push(line.replace(/\s+/g, " "));
    }
  });
  return lines.join(" ").slice(0, maxChars).trimEnd();
}

export function inferCategories(text: string): string[] {
  const terms = new Set(tokenize(text));
  const scores = Object.entries(CATEGORY_KEYWORDS)
    .map(([category, keywords]) => [category, [...terms].filter((term) => keywords.has(term)).length] as const)
    .filter(([, score]) => score > 0)
    .sort((a, b) => b[1] - a[1]);
  return scores.map(([category]) => category) || ["general"];
}

function titleFromPath(filePath: string): string {
  return path.basename(filePath, path.extname(filePath)).replace(/_/g, " ").replace(/\b\w/g, (c) => c.toUpperCase());
}
